compiler/rustc_symbol_mangling/src/typeid/typeid_itanium_cxx_abi.rs

```diff
@@ -0,0 +1,534 @@
+            for ty in tys.iter() {
+                s.push_str(&encode_ty(tcx, ty, dict, options));
+            }
+            s.push('E');
+            compress(dict, (ty, TyQ::None), &mut s);
+            typeid.push_str(&s);
+        }
+
+        ty::Array(ty0, len) => {
+            // A<array-length><element-type>
+            let mut s = String::from("A");
+            s.push_str(&format!("{}", &len.val().try_to_scalar().unwrap().to_i64().unwrap()));
+            s.push_str(&encode_ty(tcx, *ty0, dict, options));
+            compress(dict, (ty, TyQ::None), &mut s);
+            typeid.push_str(&s);
+        }
+
+        ty::Slice(ty0) => {
+            // u5sliceI<element-type>E as vendor extended type
+            let mut s = String::from("u5sliceI");
+            s.push_str(&encode_ty(tcx, *ty0, dict, options));
+            s.push('E');
+            compress(dict, (ty, TyQ::None), &mut s);
+            typeid.push_str(&s);
+        }
+
+        // User-defined types
+        ty::Adt(adt_def, substs) => {
+            let mut s = String::new();
+            let def_id = adt_def.0.did;
+            let name = format!("{}", tcx.item_name(def_id));
+            let crate_name = format!("{}", tcx.crate_name(def_id.krate));
+            // FIXME(rcvalle): Identify C char and integer type uses and encode them with their
+            // respective builtin type encoding as specified by the Itanium C++ ABI if
+            // encoding C types is enabled.
+            if name == "c_void"
+                && (crate_name == "core" || crate_name == "std" || crate_name == "libc")
+            {
+                // C char, integer, and void types
+                s.push('v');
+            } else if adt_def.repr().c() {
+                // For for cross-language CFI support, the encoding must be compatible at the FFI
+                // boundary. For instance:
+                //
+                //     struct type1 {};
+                //     void foo(struct type1* bar) {}
+                //
+                // is encoded as:
+                //
+                //     !1 = !{i64 0, !"_ZTSFvP5type1E"}
+                //
+                // So, encode any repr(C) user-defined type [or extern types (i.e., ty::Foreign)] as
+                // <length><name>, where <name> is <unscoped-name>.
+                s.push_str(&format!("{}", name.len()));
+                s.push_str(&name);
+                compress(dict, (ty, TyQ::None), &mut s);
+            } else if adt_def.is_struct() && adt_def.repr().transparent() {
+                // <element-type1>
+                let mut tys = adt_def.all_fields().map(|field| field.ty(tcx, substs));
+                let ty0 = tys.next().unwrap();
+                // Generalize any repr(transparent) user-defined type that is either a pointer or
+                // reference, and either references itself or any other type that contains or
+                // references itself, to avoid a reference cycle.
```

Comment:
> I would like to see a concrete use cases that you are trying to address with this special handling for `#[repr(transparent)]`. Examples of Rust code that would be rejected otherwise, ideally included as test cases.

Without this, for instance:

```
#[repr(transparent)]
struct Type1(i32);
fn foo(Type1) {};
fn bar(i32) {};
```

`foo` and `bar` wouldn't be compatible (i.e., `foo` couldn't be called with a `i32` argument, and `bar` couldn't be called with a `Type1` argument) with CFI enabled.

> 
> The current implementation mixes different concerns, those related to type identity and type mangling. I think those aspect should be separated. For example, the current implementation on the one hand mangles transparent types in terms of the inner type, on the other it still uses `ty::Ty` as a key which does distinguish between them, so they still generate distinct type identifiers (it does use `ty0` as key here, but consider for example a reference to this newtype, and and a reference to the inner type).
I added test cases that cover exactly this:

```
#[repr(transparent)]
struct Type10(*mut Type10);

#[repr(transparent)]
struct Type11<'a>(&'a Type11<'a>);

#[repr(transparent)]
struct Type12<'a>(&'a &'a [Type12<'a>]);
```

which are encoded as (I'll fix the dropping of type and extended type qualifiers.):

```
// CHECK: ![[TYPE37]] = !{i64 0, !"_ZTSFvPKPvE"}
// CHECK: ![[TYPE38]] = !{i64 0, !"_ZTSFu3refIu3refIvEES0_U3mutu3refIS_EE"}
// CHECK: ![[TYPE39]] = !{i64 0, !"_ZTSFu3refIu3refIvEES0_U3mutu3refIS_ES2_E"}
```
Do you have an example of a particular case you're concerned about?

> 
> Which is why I would prefer to leave this aspect for another pull request.

Sure. I don't mind sending a separate PR for this, I just want to understand the issue first.

Reply:
In terms of use cases for `#[repr(transparent)]`, what I have in mind, would examples of a real world code that would be otherwise rejected, ideally based on actual experience of running Rust code with CFI. Demonstrating that such use cases exists and are common enough that we should go out of the way to support them.

> Do you have an example of a particular case you're concerned about?

The issue I had in mind is analogous to the one with different representation void, i.e., they are only partially equivalent:

```rust
#[repr(transparent)]
pub struct S(bool);
pub fn f(_: *const bool, _: *const S) {}     // _ZTSFvPKbPKbE
pub fn g(_: *const bool, _: *const bool) {}  // _ZTSFvPKbS0_E
```

Reply:
> In terms of use cases for `#[repr(transparent)]`, what I have in mind, would examples of a real world code that would be otherwise rejected, ideally based on actual experience of running Rust code with CFI. Demonstrating that such use cases exists and are common enough that we should go out of the way to support them.

I don't think we need to find a case where CFI breaks something that uses `#[repr(transparent)]` to implement it. I think CFI shouldn't introduce any constraints that the language itself doesn't have (besides the intended purpose), which would be the case if we don't implement it.

> 
> > Do you have an example of a particular case you're concerned about?
> 
> The issue I had in mind is analogous to the one with different representation void, i.e., they are only partially equivalent:
> 
> ```rust
> #[repr(transparent)]
> pub struct S(bool);
> pub fn f(_: *const bool, _: *const S) {}     // _ZTSFvPKbPKbE
> pub fn g(_: *const bool, _: *const bool) {}  // _ZTSFvPKbS0_E
> ```

Oh, I see. You were referring to the same substitution dictionary key issue; I though I was missing something. I'll fix it soon.

Reply:
Done.